Reproducibility/src/main/java/org/tribuo/reproducibility/ReproUtil.java

```diff
@@ -151,33 +188,32 @@ private ReproUtil(ModelProvenance provenance, Model<?> originalModel) {
             }
         }
 
-        Class<? extends Trainer<T>> trainerClass = null;
-
         // Recover the name of the trainer class from the model's provenance
         // Convert to a class object so it can be passed to the config manager to recover the trainer object
         try {
-            trainerClass = (Class<? extends Trainer<T>>) Class.forName(this.modelProvenance.getTrainerProvenance().getClassName());
+            @SuppressWarnings("unchecked")
+            Class<? extends Trainer<T>> trainerClass = (Class<? extends Trainer<T>>) Class.forName(this.modelProvenance.getTrainerProvenance().getClassName());
+            // Use the class identified from the prov to create a new Trainer from the config manager that
+            // is the same as the original trainer used.
+            String trainerName = cm.listAll(trainerClass).get(0);
+            @SuppressWarnings("unchecked")
+            Trainer<T> newTrainer = (Trainer<T>) cm.lookup(trainerName);
+
+            return newTrainer;
         } catch (ClassNotFoundException e) {
             // This exception should not occur since an instance of this class is instantiated in the loop above
-            e.printStackTrace();
+            throw new IllegalStateException("Unexpected class cast exception",e);
         }
-
-        // Use the class identified from the prov to create a new Trainer from the config manager that
-        // is the same as the original trainer used.
-        String trainerName = cm.listAll(trainerClass).get(0);
-        Trainer<T> newTrainer = (Trainer<T>) cm.lookup(trainerName);
-
-        return newTrainer;
     }
 
     /**
-     * This method takes a class object and will recover that datasource from the {@link ConfigurationManager}
-     *
-     * @param dataSourceClass A configurable class that this method will search for in the configuration manager
-     * @return A {@link DataSource} object that was used to load data from te original model
-     * @throws IllegalArgumentException
+     * This method takes a class object and will recover that datasource from the {@link ConfigurationManager}.
+     * <p>
+     * Throws {@link IllegalStateException} if the data source cannot be instantiated.
+     * @param dataSourceClass A configurable class that this method will search for in the configuration manager.
+     * @return A {@link DataSource} object that was used to load data from te original model.
      */
-    private <T extends Output<T>> DataSource<T> getDatasourceFromCM(Class<? extends Configurable> dataSourceClass){
+    private DataSource<T> getDatasourceFromCM(Class<? extends Configurable> dataSourceClass){
```

Comment:
Shouldn't this be?
```suggestion
    private <D extends DataSource<T>> D getDatasourceFromCM(Class<D> dataSourceClass){
```

Reply:
I sharpened it to `<D extends ConfigurableDataSource<T>>`, as `DataSource` is not necessarily a subclass of `Configurable`.